Draft-Accepted/RFCNNNN-Install-PowerShell.md

```diff
@@ -0,0 +1,218 @@
+ 5.1 with .NET Framework provides. There is currently no plan by Windows or the .NET team to ship
+ newer .NET in Windows, so PS7 will need to continue to ship .NET with PS7.
+
+### Support lifecycle
+
+ Windows has a 5+5 year support lifecycle. .NET supports up to 3 years for Long Term Support (LTS)
+ versions. So even if size was not an issue, we cannot check in a version of PS7 with a .NET that is
+ not supported while Windows itself is supported.
+
+### Solution
+
+To solve these issues, the proposal is to ship a new cmdlet in Windows and the PowerShell Gallery
+that would;
+
+- Be discoverable using normal PowerShell conventions
+- Perform a simple download and install of the latest stable release of PowerShell
+- Provide advanced options with parameters;
+  - Specific version
+  - Stable, Preview or LTS
+  - Scope - location of installation based on permissions
+
+## Goals/Non-Goals
+
+Goals:
+
+- Be discoverable as other built-in cmdlets
+- By default, download and install the latest stable release to the users CurrentUser Scope
+  - CurrentUser - default installation scope
+  - AllUsers - Administrative users may explicit select this scope
+- Allow users to use remoting for install at scale with Invoke-Command
+- Allow advanced users to specify the version of PowerShell
+- Allow advanced users to specify scope (location of installation)
+- Allow advanced users to specify channel Stable/Preview/LTS
+- Users should receive a warning to continue the installation (ConfirmImpact=High)
+
+Non-goals:
+
+- Removal of PowerShell (Remove-PowerShell7)
+- Updating of PowerShell (Update-PowerShell7)
+- Disconnect scenarios - Users that are air-gapped may download the appropriate release from GitHub.
+- Handling of additional MSI switch options - i.e. Enable PSRemoting
+
+## Specification
+
+Cmdlet name: Install-PowerShell7
+
+Alias: None
+
+The default behavior is to download and extract the .ZIP package, then install the latest stable
+release of PowerShell into the user scope allowed by permissions. If administrative privilege,
+install to the scope of AllUsers. If user does not have administrative privilege, install to the
+scope CurrentUser.
+
+- The package install location - $env:LOCALAPPDATA\Microsoft\PowerShell
+- The installer should update the Users PATH for pwsh
+- The installer should enable Microsoft Updates for scope AllUsers. Due to permission requirements,
+  the scope CurrentUser will require manual updating.
+- Confirm impact should be set to **High** to force a warning and the ability to abort
+
+The following parameters can be added by experienced PowerShell users to customize the installation.
+
+- Parameter: Version - overrides the default and allows the user to specify the version of
+  PowerShell that will be installed. PowerShell version 7.0 and 7.1 are out of support, so version
+  numbers should begin with 7.2. This would install the latest 7.2.x.
```

Comment:
The cmdlet needs a way tell what versions are under support and what versions have reached EOL. So, when 7.2 reaches EOL, it throws error when specifying 7.2.

Reply:
This is a good point that I had not planned for.  Is there something like a tag on GitHub that we can check?

Reply:
I don't think currently there is a way to tell this programmatically. How is this tracked by our docs?

Reply:
Manual effort

Reply:
Today, Travis provides a micro-service that helps us update EOL.  Install-PowerShell could consume this information. The cmdlet can provide a warning or error if a version is out of support.  The bigger question is if we support -version (for Ci scenarios) then we would need to support version that are EOL.  If we don't believe this cmdlet should be used for ci scenarios, then we could remove -version and not provide unsupported version.  We could try to be more like docker, 30 days before the EOL, users would start getting a warning, after EOL, the image will be unavailable and user would receive an error.  This brings us back to a scope discussion - this cmdlet was intended for Windows users to upgrade to the latest stable - I think that is equivalent to WinGet.  I suggest we don't exceed the capabilities for WinGet, which means not supporting ci scenarios and removing the -version parameter.  I recommend at minimum the LTS, Stable, preview

Reply:
After having several discussions - focusing the cmdlet on a simple installation of Stable, Preview and LTS only.  Removing Version.  The cmdlet should install only the latest available.